implementations/rust/ockam/ockam_identity/src/secure_channel/decryptor.rs

```diff
@@ -1,39 +1,137 @@
 use crate::identity::IdentityError;
-use crate::secure_channel::encryptor::Encryptor;
+use crate::secure_channel::encryptor::{Encryptor, KEY_RENEWAL_INTERVAL};
 use ockam_core::compat::sync::Arc;
 use ockam_core::compat::vec::Vec;
-use ockam_core::vault::{KeyId, SymmetricVault};
+use ockam_core::vault::KeyId;
 use ockam_core::Result;
+use ockam_key_exchange_xx::XXInitializedVault;
+use tracing::warn;
 
-#[derive(Clone)]
 pub(crate) struct Decryptor {
-    key: KeyId,
-    vault: Arc<dyn SymmetricVault>,
+    current_key: KeyId,
+    current_key_nonce: u64,
+
+    previous_key: Option<KeyId>,
+    vault: Arc<dyn XXInitializedVault>,
+    nonce_tracker: NonceTracker,
+}
+
+pub(crate) struct NonceTracker {
+    nonce_bitmap: u16,
+    current_nonce: u64,
 }
 
 impl Decryptor {
     /// Restore 12-byte nonce needed for AES GCM from 8 byte that we use for noise
-    fn convert_nonce_from_small(b: &[u8]) -> Result<[u8; 12]> {
+    fn convert_nonce_from_small(b: &[u8]) -> Result<(u64, [u8; 12])> {
         let bytes: [u8; 8] = b.try_into().map_err(|_| IdentityError::InvalidNonce)?;
 
         let nonce = u64::from_be_bytes(bytes);
 
-        Ok(Encryptor::convert_nonce_from_u64(nonce).1)
+        Ok((nonce, Encryptor::convert_nonce_from_u64(nonce).1))
     }
 
-    pub async fn decrypt(&self, payload: &[u8]) -> Result<Vec<u8>> {
+    pub async fn decrypt(&mut self, payload: &[u8]) -> Result<Vec<u8>> {
         if payload.len() < 8 {
             return Err(IdentityError::InvalidNonce.into());
         }
 
-        let nonce = Self::convert_nonce_from_small(&payload[..8])?;
+        let (nonce, nonce_buffer) = Self::convert_nonce_from_small(&payload[..8])?;
+
+        self.nonce_tracker.mark(nonce)?;
```

Comment:
the actual marking should be done only after decryption succeed

Reply:
Makes sense

Reply:
btw there is a test case testing this https://github.com/build-trust/ockam/blob/65979929bf310f84c62df2fe997b908987db077f/implementations/rust/ockam/ockam_identity/src/secure_channel/mod.rs#L150-L186